Using a Processor makes rendering documents with a TOC fail

As first observed in asciidoctor/asciidoctor-maven-plugin#258 when using a Processor that iterates over the blocks of a document, rendering that Document with a table of contents will fail.

That means with this Processor:
```java
public class MyTreeprocessor extends Treeprocessor {
    public TerminalCommandTreeprocessor(Map<String, Object> config) {
        super(config);
    }

    @Override
    public Document process(Document document) {
        final List<AbstractBlock> blocks = this.document.blocks();
        return document;
    }
```
this code will fail:
```java
        JavaExtensionRegistry javaExtensionRegistry = this.asciidoctor.javaExtensionRegistry();
        javaExtensionRegistry.treeprocessor(MyTreeprocessor.class);
        String content = asciidoctor.renderFile(
            classpath.getResource("sample-with-sections.ad"),
            options().toFile(false)
                .attributes(AttributesBuilder.attributes().tableOfContents(true))
                .get());
```

The exception is this:
```
org.asciidoctor.internal.AsciidoctorCoreException: org.jruby.exceptions.RaiseException: (NoMethodError) undefined method `level' for nil:NilClass
	at org.asciidoctor.internal.JRubyAsciidoctor.renderFile(JRubyAsciidoctor.java:345)
	at org.asciidoctor.internal.JRubyAsciidoctor.renderFile(JRubyAsciidoctor.java:434)
	at org.asciidoctor.extension.WhenJavaExtensionIsRegistered.when_using_a_tree_processor_a_toc_should_still_be_created(WhenJavaExtensionIsRegistered.java:401)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.asciidoctor.util.ClasspathResources$1.evaluate(ClasspathResources.java:24)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
Caused by: org.jruby.exceptions.RaiseException: (NoMethodError) undefined method `level' for nil:NilClass
	at RUBY.outline(/Users/robertpanzer/dev/asciidoctorj/asciidoctorj-core/build/resources/main/gems/asciidoctor-1.5.4/lib/asciidoctor/converter/html5.rb:300)
	at RUBY.embedded(/Users/robertpanzer/dev/asciidoctorj/asciidoctorj-core/build/resources/main/gems/asciidoctor-1.5.4/lib/asciidoctor/converter/html5.rb:272)
	at RUBY.convert(/Users/robertpanzer/dev/asciidoctorj/asciidoctorj-core/build/resources/main/gems/asciidoctor-1.5.4/lib/asciidoctor/converter/base.rb:34)
	at RUBY.convert(/Users/robertpanzer/dev/asciidoctorj/asciidoctorj-core/build/resources/main/gems/asciidoctor-1.5.4/lib/asciidoctor/document.rb:1044)
	at RUBY.convert(/Users/robertpanzer/dev/asciidoctorj/asciidoctorj-core/build/resources/main/gems/asciidoctor-1.5.4/lib/asciidoctor.rb:1503)
	at RUBY.convert_file(/Users/robertpanzer/dev/asciidoctorj/asciidoctorj-core/build/resources/main/gems/asciidoctor-1.5.4/lib/asciidoctor.rb:1576)
	at RUBY.convertFile(<script>:68)
	at org.jruby.gen.InterfaceImpl1854727011.convertFile(org/jruby/gen/InterfaceImpl1854727011.gen:13)
```

--- a/asciidoctorj-core/build.gradle
+++ b/asciidoctorj-core/build.gradle
@@ -27,6 +27,14 @@ def gemFiles = fileTree(jruby.gemInstallDir) {
 jrubyPrepareGems << {
   copy { // bundles the gems inside this artifact
     from gemFiles
+    eachFile {
+      // See https://github.com/asciidoctor/asciidoctorj/issues/497
+      if (it.path ==~ /gems\/asciidoctor-.+\/lib\/asciidoctor\/abstract_block.rb/) {
+        it.filter { line ->
+          line.replaceAll(/block.context == :section/, 'block.context.to_sym == :section')
+        }
+      }
+    }
     into sourceSets.main.output.resourcesDir
   }
 }

--- a/asciidoctorj-core/src/main/java/org/asciidoctor/ast/AbstractBlockImpl.java
+++ b/asciidoctorj-core/src/main/java/org/asciidoctor/ast/AbstractBlockImpl.java
@@ -1,15 +1,13 @@
 package org.asciidoctor.ast;
 
-import java.util.List;
-import java.util.Map;
-
-import org.asciidoctor.converter.ConverterProxy;
 import org.asciidoctor.internal.RubyHashUtil;
 import org.asciidoctor.internal.RubyUtils;
 import org.jruby.Ruby;
 import org.jruby.RubyArray;
 import org.jruby.RubyObject;
-import org.jruby.javasupport.JavaEmbedUtils;
+
+import java.util.List;
+import java.util.Map;
 
 public class AbstractBlockImpl extends AbstractNodeImpl implements AbstractBlock {
 
@@ -42,6 +40,10 @@ public class AbstractBlockImpl extends AbstractNodeImpl implements AbstractBlock
     @Override
     public String getStyle() {
         return delegate.getStyle();
+    }
+
+    public String getCaption() {
+        return RubyUtils.invokeRubyMethod(delegate, "caption", new Object[0], String.class);
     }
 
     @Override

--- a/asciidoctorj-core/src/main/java/org/asciidoctor/ast/SectionImpl.java
+++ b/asciidoctorj-core/src/main/java/org/asciidoctor/ast/SectionImpl.java
@@ -1,6 +1,9 @@
 package org.asciidoctor.ast;
 
+import org.asciidoctor.internal.RubyUtils;
 import org.jruby.Ruby;
+
+import java.util.List;
 
 public class SectionImpl extends AbstractBlockImpl implements Section {
 
@@ -36,4 +39,27 @@ public class SectionImpl extends AbstractBlockImpl implements Section {
         return this.delegate.number();
     }
 
+    public String sectnum() {
+        return RubyUtils.invokeRubyMethod(delegate, "sectnum", new Object[0], String.class);
+    }
+
+    public String sectnum(String delimiter) {
+        return RubyUtils.invokeRubyMethod(delegate, "sectnum", new Object[]{delimiter}, String.class);
+    }
+
+    public String sectnum(String delimiter, boolean append) {
+        return RubyUtils.invokeRubyMethod(delegate, "sectnum", new Object[]{delimiter, append}, String.class);
+    }
+
+    public List<Section> getSections() {
+        return RubyUtils.invokeRubyMethod(delegate, "sections", new Object[0], List.class);
+    }
+
+    public boolean isSections() {
+        return RubyUtils.invokeRubyMethod(delegate, "sections?", new Object[0], Boolean.class);
+    }
+
+    public String getCaptionedTitle() {
+        return RubyUtils.invokeRubyMethod(delegate, "captioned_title", new Object[0], String.class);
+    }
 }

--- a/asciidoctorj-core/src/main/java/org/asciidoctor/internal/RubyUtils.java
+++ b/asciidoctorj-core/src/main/java/org/asciidoctor/internal/RubyUtils.java
@@ -7,6 +7,7 @@ import org.jruby.RubyClass;
 import org.jruby.RubyObject;
 import org.jruby.RubySymbol;
 import org.jruby.internal.runtime.GlobalVariable.Scope;
+import org.jruby.java.proxies.RubyObjectHolderProxy;
 import org.jruby.javasupport.JavaClass;
 import org.jruby.javasupport.JavaEmbedUtils;
 import org.jruby.runtime.GlobalVariable;
@@ -39,6 +40,19 @@ public class RubyUtils {
     public static final void setGlobalVariable(Ruby rubyRuntime, String variableName, Object variableValue) {
         String script = String.format("$%s = %s", variableName, variableValue);
         rubyRuntime.evalScriptlet(script);
+    }
+
+    public static <T> T invokeRubyMethod(final Object target, final String methodName, final Object[] args, final Class<T> resultType) {
+        IRubyObject rubyObject = null;
+        if (target instanceof RubyObjectHolderProxy) {
+            rubyObject = RubyObjectHolderProxy.class.cast(target).__ruby_object();
+        } else if (target instanceof IRubyObject) {
+            rubyObject = (IRubyObject) target;
+        }
+        if (rubyObject != null) {
+            return (T) JavaEmbedUtils.invokeMethod(rubyObject.getRuntime(), rubyObject, methodName, args, resultType);
+        }
+        throw new IllegalArgumentException("Target is a " + target.getClass() + " instead of a Ruby object.");
     }
 
 }

--- a/asciidoctorj-core/src/test/java/org/asciidoctor/extension/WhenJavaExtensionIsRegistered.java
+++ b/asciidoctorj-core/src/test/java/org/asciidoctor/extension/WhenJavaExtensionIsRegistered.java
@@ -22,6 +22,7 @@ import java.util.HashMap;
 import java.util.Map;
 
 import org.asciidoctor.Asciidoctor;
+import org.asciidoctor.AttributesBuilder;
 import org.asciidoctor.Options;
 import org.asciidoctor.SafeMode;
 import org.asciidoctor.ast.DocumentRuby;
@@ -389,6 +390,24 @@ public class WhenJavaExtensionIsRegistered {
         assertThat(contentElement.text(), is("gem install asciidoctor"));
 
     }
+
+    /**
+     * See https://github.com/asciidoctor/asciidoctorj/issues/497.
+     */
+    @Test
+    public void when_using_a_tree_processor_a_toc_should_still_be_created() {
+
+        JavaExtensionRegistry javaExtensionRegistry = this.asciidoctor.javaExtensionRegistry();
+
+        javaExtensionRegistry.treeprocessor(TerminalCommandTreeprocessor.class);
+
+        String content = asciidoctor.renderFile(
+            classpath.getResource("sample-with-sections.ad"),
+            options().toFile(false)
+                .attributes(AttributesBuilder.attributes().tableOfContents(true))
+                .get());
+    }
+
 
     @Test
     public void a_treeprocessor_as_string_should_be_executed_in_document() {

--- a/asciidoctorj-core/src/test/resources/sample-with-sections.ad
+++ b/asciidoctorj-core/src/test/resources/sample-with-sections.ad
@@ -1,0 +1,15 @@
+= Hello World
+
+Some text
+
+== Section A
+
+More text
+
+=== Subsection A1
+
+Subsections are also so important.
+
+== Section B
+
+And even more text

